src/solc_0.8/ReferralValidator/ReferralValidator08.sol

```diff
@@ -0,0 +1,186 @@
+
+    event ReferralUsed(
+        address indexed referrer,
+        address indexed referee,
+        address indexed token,
+        uint256 amount,
+        uint256 commission,
+        uint256 commissionRate
+    );
+
+    constructor(
+        address initialSigningWallet,
+        uint256 initialMaxCommissionRate,
+        address admin
+    ) {
+        _signingWallet = initialSigningWallet;
+        _maxCommissionRate = initialMaxCommissionRate;
+        _admin = admin;
+    }
+
+    /**
+     * @notice Update the signing wallet
+     * @param newSigningWallet The new address of the signing wallet
+     */
+    function updateSigningWallet(address newSigningWallet) external {
+        require(_admin == msg.sender, "Sender not admin");
+        _previousSigningWallets[_signingWallet] = block.timestamp + _previousSigningDelay;
+        _signingWallet = newSigningWallet;
+    }
+
+    /**
+     * @dev signing wallet authorized for referral
+     * @return the address of the signing wallet
+     */
+    function getSigningWallet() external view returns (address) {
+        return _signingWallet;
+    }
+
+    /**
+     * @dev Update the maximum commission rate
+     * @param newMaxCommissionRate The new maximum commission rate
+     */
+    function updateMaxCommissionRate(uint256 newMaxCommissionRate) external {
+        require(_admin == msg.sender, "Sender not admin");
+        _maxCommissionRate = newMaxCommissionRate;
+    }
+
+    /**
+     * @notice the max commision rate
+     * @return the maximum commision rate that a referral can give
+     */
+    function getMaxCommisionRate() external view returns (uint256) {
+        return _maxCommissionRate;
+    }
+
+    function handleReferralWithETH(
+        uint256 amount,
+        bytes memory referral,
+        address payable destination
+    ) internal {
+        uint256 amountForDestination = amount;
+
+        if (referral.length > 0) {
+            (bytes memory signature, address referrer, address referee, uint256 expiryTime, uint256 commissionRate) =
+                decodeReferral(referral);
+
+            uint256 commission = 0;
+
+            if (isReferralValid(signature, referrer, referee, expiryTime, commissionRate)) {
+                commission = SafeMath.div(SafeMath.mul(amount, commissionRate), 10000);
+
+                emit ReferralUsed(referrer, referee, address(0), amount, commission, commissionRate);
+                amountForDestination = SafeMath.sub(amountForDestination, commission);
+            }
+
+            if (commission > 0) {
+                payable(referrer).transfer(commission);
+            }
+        }
+
+        destination.transfer(amountForDestination);
+    }
+
+    function handleReferralWithERC20(
+        address buyer,
+        uint256 amount,
+        bytes memory referral,
+        address payable destination,
+        address tokenAddress
+    ) internal {
+        ERC20Token token = ERC20Token(tokenAddress);
```

Comment:
Use the ERC20 interface from openzeppelin

Reply:
fixed -> c21e258